scripts/toltec/recipe.py

```diff
@@ -0,0 +1,225 @@
+                    pkg_name, self, functions[pkg_name]
+                )
+
+    @classmethod
+    def from_file(cls, name: str, root: str) -> "Recipe":
+        """Load a recipe from a file."""
+        with open(os.path.join(root, "package"), "r") as recipe:
+            return Recipe(name, root, recipe.read())
+
+
+class Package:  # pylint:disable=too-many-instance-attributes
+    """Load and execute a package from a recipe."""
+
+    def __init__(self, name: str, parent: Recipe, source: str):
+        """
+        Load a package from a Bash source.
+
+        :param name: name of the package
+        :param parent: recipe which declares this package
+        :param source: source string of the package (either the full recipe
+            script if it contains only a single package, or the package
+            script for split packages)
+        :raises RecipeError: if the package contains an error
+        """
+        self.name = name
+        self.parent = parent
+        variables, functions = bash.get_declarations(source)
+        self.variables = variables = {**parent.variables, **variables}
+        self.functions = functions = {**parent.functions, **functions}
+
+        # Parse and check package metadata
+        pkgver_str = _check_field_string(variables, "pkgver")
+        self.pkgver = version.Version.parse(pkgver_str)
+
+        self.arch = _check_field_string(variables, "arch", "armv7-3.2")
+        self.pkgdesc = _check_field_string(variables, "pkgdesc")
+        self.url = _check_field_string(variables, "url")
+        self.section = _check_field_string(variables, "section")
+        self.license = _check_field_string(variables, "license")
+        self.depends = _check_field_indexed(variables, "depends", [])
+        self.conflicts = _check_field_indexed(variables, "conflicts", [])
+
+        if "package" not in functions:
+            raise RecipeError(
+                "Missing required function package() \
+for package {self.name}"
+            )
+
+        self.action = functions["package"]
+        self.install = {}
+
+        for action in ("preinstall", "configure"):
+            self.install[action] = functions.get(action, "")
+
+        for rel, step in product(("pre", "post"), ("remove", "upgrade")):
+            self.install[rel + step] = functions.get(rel + step, "")
+
+    def pkgid(self) -> str:
+        """Get the unique identifier of this package."""
+        return "_".join((self.name, str(self.pkgver), self.arch))
+
+    def filename(self) -> str:
+        """Get the name of the archive corresponding to this package."""
+        return self.pkgid() + ".ipk"
+
+    def control_fields(self) -> str:
+        """Get the control fields for this package."""
+        control = f"""Package: {self.name}
+Version: {self.pkgver}
+Maintainer: {self.parent.maintainer}
+Section: {self.section}
+Architecture: {self.arch}
+Description: {self.pkgdesc}
+HomePage: {self.url}
+License: {self.license}
+"""
```

Comment:
```suggestion
        control = inspect.cleandoc(f"""
            Package: {self.name}
            Version: {self.pkgver}
            Maintainer: {self.parent.maintainer}
            Section: {self.section}
            Architecture: {self.arch}
            Description: {self.pkgdesc}
            HomePage: {self.url}
            License: {self.license}
            """)
```

(need to add `inspect` to the imports for it to work!)

Not sure if wanted, but can get indented better and still get almost the same result (no trailing newline) with `inspect.cleandoc`. Alternatively `textwrap.dedent` can be used as well but this wont remove the first newline here (maybe add lstrip() as well to remove it).

Reply:
Nice, I’ll use that, thanks! This can also be useful in `Builder._archive`.